### PR TITLE
Heap Problems

### PR DESCRIPTION
**Pull request by AHarishma**
Merge K Sorted List

### CODE REVIEW

File: src/test/java/com/company/heap/MergeKSortedListTest.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+public class MergeKSortedListTest {
+    @ParameterizedTest
+    @MethodSource("testDataProvider")
+    void testMergeKSortedList(ListNode[] lists, ListNode expectedResult) {
+        ListNode result = new MergeKSortedList().mergeKLists(lists);
+        assertTrue(ListNode.areEqual(expectedResult, result));
+    }
+
+    static Stream<Arguments> testDataProvider() {
```
Comment (madhusdhnn) [inline src/test/java/com/company/heap/MergeKSortedListTest.java]: could be private

File: src/test/java/com/company/heap/MergeKSortedListTest.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+import org.junit.jupiter.params.provider.MethodSource;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+import java.util.stream.Stream;
+
+public class MergeKSortedListTest {
+    @ParameterizedTest
+    @MethodSource("testDataProvider")
+    void testMergeKSortedList(ListNode[] lists, ListNode expectedResult) {
```
Comment (madhusdhnn) [inline src/test/java/com/company/heap/MergeKSortedListTest.java]: don't forget access modifiers (private, public, etc)

File: src/main/java/com/company/linkedlist/ListNode.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    public ListNode(int val) {
+        this.val = val;
+    }
+
+    public ListNode(int val, ListNode next) {
+        this.val = val;
+        this.next = next;
+    }
+
+    public static boolean areEqual(ListNode head1, ListNode head2) {
```
Comment (madhusdhnn) [inline src/main/java/com/company/linkedlist/ListNode.java]: You can move this to the test function as you are using it for comparing elements in the test

**Review by madhusdhnn — CHANGES_REQUESTED** [review]
Complete the changes requested and share the PR again

File: src/main/java/com/company/linkedlist/ListNode.java (hunk 3 above)
Comment (AHarishma) [reply src/main/java/com/company/linkedlist/ListNode.java]: here we are comparing two list node can we not keep in ListNode class

File: src/test/java/com/company/heap/MergeKSortedListTest.java (hunk 1 above)
Comment (AHarishma) [reply src/test/java/com/company/heap/MergeKSortedListTest.java]: changed

File: src/test/java/com/company/heap/MergeKSortedListTest.java (hunk 2 above)
Comment (AHarishma) [reply src/test/java/com/company/heap/MergeKSortedListTest.java]: sure added

File: src/main/java/com.company/heap/RunningMedian.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+            } else {
+                minHeap.add(arr[i]);
+            }
+            if (minHeap.size() - maxHeap.size() > 1) {
+                maxHeap.add(minHeap.poll());
+            }
+            if (maxHeap.size() - minHeap.size() > 1) {
+                minHeap.add(maxHeap.poll());
+            }
+            if (maxHeap.size() == minHeap.size()) {
```
Comment (madhusdhnn) [inline src/main/java/com.company/heap/RunningMedian.java]: You could combine > and = in single if condition ( >= )

File: src/main/java/com/company/linkedlist/ListNode.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
+package com.company.linkedlist;
+
+public class ListNode {
+    public int val;
+    public ListNode next;
+
+    public ListNode() {
```
Comment (madhusdhnn) [inline src/main/java/com/company/linkedlist/ListNode.java]: Use Lombok annotations

File: src/main/java/com/company/linkedlist/ListNode.java (hunk 3 above)
Comment (madhusdhnn) [reply src/main/java/com/company/linkedlist/ListNode.java]: Ok. Remove static in that case and keep it in the instance level. Accept only one parameter which is the head of the other list. first list will be `this`.

File: src/test/java/com/company/heap/RunningMedianTest.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+import java.util.stream.Stream;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class RunningMedianTest {
+    @ParameterizedTest
+    @MethodSource("testDataProvider")
+    public void testFindRunningMedian(int[] arr, int[] expectedResult) {
+        int[] result = new RunningMedian().findRunningMedian(arr);
+        assertEquals(Arrays.toString(expectedResult), Arrays.toString(result));
```
Comment (madhusdhnn) [inline src/test/java/com/company/heap/RunningMedianTest.java]: You can use `assertArrayEquals` for asserting arrays

**Review by madhusdhnn — CHANGES_REQUESTED** [review]

File: src/main/java/com.company/heap/RunningMedian.java (hunk 4 above)
Comment (AHarishma) [reply src/main/java/com.company/heap/RunningMedian.java]: changed

File: src/test/java/com/company/heap/RunningMedianTest.java (hunk 6 above)
Comment (AHarishma) [reply src/test/java/com/company/heap/RunningMedianTest.java]: changed

File: src/main/java/com/company/linkedlist/ListNode.java (hunk 5 above)
Comment (AHarishma) [reply src/main/java/com/company/linkedlist/ListNode.java]: changed

File: src/main/java/com/company/linkedlist/ListNode.java (hunk 3 above)
Comment (AHarishma) [reply src/main/java/com/company/linkedlist/ListNode.java]: getting null pointer exception

not sure how to send linked list [[]] and []

File: src/main/java/com.company/heap/RunningMedian.java [hunk 7]
```diff
@@ -0,0 +1,101 @@
+package com.company.heap;
```
Comment (madhusdhnn) [inline src/main/java/com.company/heap/RunningMedian.java:1]: Make sure to solve this in Leetcode and GFG also

File: src/main/java/com/company/linkedlist/ListNode.java (hunk 3 above)
Comment (madhusdhnn) [reply src/main/java/com/company/linkedlist/ListNode.java]: Why one is `[[]]` and the other is `[]`? Both should be in the same depth only. If it is different, move this method some util class like `LinkedListUtils`

**Review by madhusdhnn — CHANGES_REQUESTED** [review]

File: src/main/java/com.company/heap/RunningMedian.java (hunk 7 above)
Comment (AHarishma) [reply src/main/java/com.company/heap/RunningMedian.java:1]: yes solved

File: src/main/java/com/company/linkedlist/ListNode.java (hunk 3 above)
Comment (AHarishma) [reply src/main/java/com/company/linkedlist/ListNode.java]: Added LinkedListUtils

File: src/main/java/com/company/linkedlist/LinkedListUtils.java [hunk 8]
```diff
@@ -0,0 +1,14 @@
+package com.company.linkedlist;
+
+public class LinkedListUtils {
+    public static boolean areEqual(ListNode head1, ListNode head2) {
+        while (head1 != null && head2 != null) {
+            if (head1.val != head2.val) {
```
Comment (madhusdhnn) [inline src/main/java/com/company/linkedlist/LinkedListUtils.java:6]: You have to check for both value and reference is same or not.

**Review by madhusdhnn — CHANGES_REQUESTED** [review]

**Pull request merged**